commons/zenoh-util/src/keyexpr_tree/arc_tree.rs

```diff
@@ -0,0 +1,617 @@
+        } else {
+            IterOrOption::Opt(self.node_mut(token, key))
+        }
+    }
+
+    type InclusionItem = Self::Node;
+    type Inclusion = IterOrOption<
+        TokenPacker<
+            Inclusion<
+                'a,
+                Children,
+                Arc<TokenCell<KeArcTreeNode<Weight, Weak<()>, Wildness, Children, Token>, Token>>,
+                Weight,
+            >,
+            &'a Token,
+        >,
+        Self::InclusionItem,
+    >;
+    fn included_nodes(&'a self, token: &'a Token, key: &'a keyexpr) -> Self::Inclusion {
+        let Ok(inner) = self.inner.try_borrow(token) else {panic!("Attempted to use KeArcTree with the wrong Token")};
+        if inner.wildness.get() {
+            IterOrOption::Iter(TokenPacker {
+                iter: Inclusion::new(&inner.children, key),
+                token,
+            })
+        } else {
+            IterOrOption::Opt(self.node(token, key))
+        }
+    }
+    type InclusionItemMut = Self::NodeMut;
+    type InclusionMut = IterOrOption<
+        TokenPacker<
+            Inclusion<
+                'a,
+                Children,
+                Arc<TokenCell<KeArcTreeNode<Weight, Weak<()>, Wildness, Children, Token>, Token>>,
+                Weight,
+            >,
+            &'a mut Token,
+        >,
+        Self::InclusionItemMut,
+    >;
+    fn included_nodes_mut(&'a self, token: &'a mut Token, key: &'a keyexpr) -> Self::InclusionMut {
+        let Ok(inner) = self.inner.try_borrow(token) else {panic!("Attempted to use KeArcTree with the wrong Token")};
+        if inner.wildness.get() {
+            IterOrOption::Iter(TokenPacker {
+                iter: Inclusion::new(unsafe { core::mem::transmute(&inner.children) }, key),
+                token,
+            })
+        } else {
+            IterOrOption::Opt(self.node_mut(token, key))
+        }
+    }
+    type PruneNode = KeArcTreeNode<Weight, Weak<()>, Wildness, Children, Token>;
+
+    fn prune_where<F: FnMut(&mut Self::PruneNode) -> bool>(
+        &self,
+        token: &mut Token,
+        mut predicate: F,
+    ) {
+        let mut wild = false;
+        let Ok(inner) = self.inner.try_borrow_mut(token) else {panic!("Attempted to use KeArcTree with the wrong Token")};
+        inner.children.filter_out(&mut |child| match unsafe {
+            (*child.get())._prune(&mut predicate)
+        } {
+            PruneResult::Delete => Arc::strong_count(child) <= 1,
+            PruneResult::NonWild => false,
+            PruneResult::Wild => {
+                wild = true;
+                false
+            }
+        });
+        inner.wildness.set(wild);
+    }
+}
+pub struct TokenPacker<I, T> {
+    iter: I,
+    token: T,
+}
+
+impl<I: Iterator, T> Iterator for TokenPacker<I, T> {
+    type Item = (I::Item, T);
+    fn next(&mut self) -> Option<Self::Item> {
+        self.iter
+            .next()
+            .map(|i| (i, unsafe { core::mem::transmute_copy(&self.token) }))
```

Comment:
Here, I wish `UnsafeIterator` or better yet `Iterator::Item<'a>` was a thing, because this is always:
- Safe when `self.token` is typed `&Token`, since we can copy these references all we want
- Very unsafe when `self.token` is typed `&mut Token`, because it becomes a mutable alias factory if these copies of the token are retained between iterations.

However, I consider the alternatives worse:
- Iterators that only yield `&TokenCell` would allow mutation, but user would have to unsafely create a `(&Token, &mut Token)` pair anyway, as `&Token` is necessary to iterate.
- Returning a `(Iterator, &mut Token)` tuple is slightly less unsafe (it can still be used to modify the tree while holding on to the iterator, which would be very bad), but is much less comfortable to use. It also loses the ability to look at multiple nodes at the same time after iterating.
- I'll try exploring returning guards built from the currently returned tuple that don't let the user extract the tokens, preventing any potential token proliferation

Reply:
Sealed guards implemented, a better solution overall :)